lib/Driver/ToolChains/Flang.cpp

```diff
@@ -294,28 +294,34 @@ void FlangFrontend::ConstructJob(Compilation &C, const JobAction &JA,
     CommonCmdArgs.push_back("8");
   }
 
-  // -g should produce DWARFv2
-  for (auto Arg : Args.filtered(options::OPT_g_Flag)) {
-    Arg->claim();
-    CommonCmdArgs.push_back("-x");
-    CommonCmdArgs.push_back("120");
-    CommonCmdArgs.push_back("0x200");
-  }
+  // Last argument of -g/-gdwarfX should be taken.
+  Arg *GArg = Args.getLastArg(options::OPT_g_Flag);
+  Arg *GDwarfArg = Args.getLastArg(options::OPT_gdwarf_2,
+                                   options::OPT_gdwarf_3,
+                                   options::OPT_gdwarf_4,
+                                   options::OPT_gdwarf_5);
+
+  if (GArg || GDwarfArg) {
+
+    for (auto Arg : Args.filtered(options::OPT_g_Flag, options::OPT_gdwarf_2,
+                                  options::OPT_gdwarf_3, options::OPT_gdwarf_4,
+                                  options::OPT_gdwarf_5)) {
+      Arg->claim();
+    }
 
-  // -gdwarf-2
-  for (auto Arg : Args.filtered(options::OPT_gdwarf_2)) {
-    Arg->claim();
     CommonCmdArgs.push_back("-x");
     CommonCmdArgs.push_back("120");
-    CommonCmdArgs.push_back("0x200");
-  }
 
-  // -gdwarf-3
-  for (auto Arg : Args.filtered(options::OPT_gdwarf_3)) {
-    Arg->claim();
-    CommonCmdArgs.push_back("-x");
-    CommonCmdArgs.push_back("120");
-    CommonCmdArgs.push_back("0x4000");
+    if (!GDwarfArg) // -g should produce default (DWARFv2)
+      CommonCmdArgs.push_back("0x200");
```

Comment:
Do we document this as the default anywhere?

Reply:
The current fix does not change the behavior of dwarf version chosen when only -g is mentioned. But you are right DWARF version 2 looks a bit awkward. Another fix I planned to submit it about this. Below are the details.

Title: Default dwarf version should be 4 for -g with flang
Details:

Currently flang dumps dwarf version 2 for -g and 4 for absence of -g (debug lite)
0x00000000: Compile Unit: length = 0x0000016c version = 0x0002 abbr_offset = 0x0000 addr_size = 0x08 (next unit at 0x00000170)

0x00000000: Compile Unit: length = 0x00000088 version = 0x0004 abbr_offset = 0x0000 addr_size = 0x08 (next unit at 0x0000008c)

It should be 4 for -g as it is the case with clang.

Reply:
Would it not be a good idea to save a step and make DWARF 4 the default here? As you point out, that would follow clang's lead. (After all, DWARF 2 is a 27 year old standard now.)

Reply:
Yes, it makes sense to club it here itself due to quantity of code-change, but I am in doubt due to impact of it being on behavior of compiler (better debugging with default option) and this suggests an individual check-in for changing default to 4. Please guide me.

Reply:
For Flang, it seems to me pretty safe to move the default to DWARF 4.

Reply:
In the summary of this change, there're examples seem to be misleading,

"-gdwarf-5 -gdwarf-3 => flang chooses 3 while clang/gfortran choose 5
-gdwarf-5 -g => flang choses the default while clang/gfortran choose 5"

Clang/gfortran choose the last, shouldn't "-gdwarf-5" appear as the last in the above examples?